File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcProviderPackageCatalog.cs
namespace Sdkwork.Rtc.Sdk;

using System.Linq;

public sealed record RtcProviderPackageCatalogEntry(
    string providerKey,
    string pluginId,
    string driverId,
    string packageIdentity,
    string manifestPath,
    string readmePath,
    string sourcePath,
    string sourceSymbol,
    bool builtin,
    bool rootPublic,
    string status,
    string runtimeBridgeStatus
);

public static class RtcProviderPackageCatalog
{
    public static readonly IReadOnlyList<RtcProviderPackageCatalogEntry> Entries =
    [
        new("volcengine", "rtc-volcengine", "sdkwork-rtc-driver-volcengine", "Sdkwork.Rtc.Sdk.Provider.Volcengine", "providers/Sdkwork.Rtc.Sdk.Provider.Volcengine/Sdkwork.Rtc.Sdk.Provider.Volcengine.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Volcengine/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Volcengine/src/RtcProviderVolcenginePackageContract.cs", "RtcProviderVolcenginePackageContract", true, false, "future-runtime-bridge-only", "reserved"),
        new("aliyun", "rtc-aliyun", "sdkwork-rtc-driver-aliyun", "Sdkwork.Rtc.Sdk.Provider.Aliyun", "providers/Sdkwork.Rtc.Sdk.Provider.Aliyun/Sdkwork.Rtc.Sdk.Provider.Aliyun.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Aliyun/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Aliyun/src/RtcProviderAliyunPackageContract.cs", "RtcProviderAliyunPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
        new("tencent", "rtc-tencent", "sdkwork-rtc-driver-tencent", "Sdkwork.Rtc.Sdk.Provider.Tencent", "providers/Sdkwork.Rtc.Sdk.Provider.Tencent/Sdkwork.Rtc.Sdk.Provider.Tencent.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Tencent/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Tencent/src/RtcProviderTencentPackageContract.cs", "RtcProviderTencentPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
        new("agora", "rtc-agora", "sdkwork-rtc-driver-agora", "Sdkwork.Rtc.Sdk.Provider.Agora", "providers/Sdkwork.Rtc.Sdk.Provider.Agora/Sdkwork.Rtc.Sdk.Provider.Agora.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Agora/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Agora/src/RtcProviderAgoraPackageContract.cs", "RtcProviderAgoraPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("zego", "rtc-zego", "sdkwork-rtc-driver-zego", "Sdkwork.Rtc.Sdk.Provider.Zego", "providers/Sdkwork.Rtc.Sdk.Provider.Zego/Sdkwork.Rtc.Sdk.Provider.Zego.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Zego/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Zego/src/RtcProviderZegoPackageContract.cs", "RtcProviderZegoPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("livekit", "rtc-livekit", "sdkwork-rtc-driver-livekit", "Sdkwork.Rtc.Sdk.Provider.Livekit", "providers/Sdkwork.Rtc.Sdk.Provider.Livekit/Sdkwork.Rtc.Sdk.Provider.Livekit.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Livekit/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Livekit/src/RtcProviderLivekitPackageContract.cs", "RtcProviderLivekitPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("twilio", "rtc-twilio", "sdkwork-rtc-driver-twilio", "Sdkwork.Rtc.Sdk.Provider.Twilio", "providers/Sdkwork.Rtc.Sdk.Provider.Twilio/Sdkwork.Rtc.Sdk.Provider.Twilio.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Twilio/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Twilio/src/RtcProviderTwilioPackageContract.cs", "RtcProviderTwilioPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("jitsi", "rtc-jitsi", "sdkwork-rtc-driver-jitsi", "Sdkwork.Rtc.Sdk.Provider.Jitsi", "providers/Sdkwork.Rtc.Sdk.Provider.Jitsi/Sdkwork.Rtc.Sdk.Provider.Jitsi.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Jitsi/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Jitsi/src/RtcProviderJitsiPackageContract.cs", "RtcProviderJitsiPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("janus", "rtc-janus", "sdkwork-rtc-driver-janus", "Sdkwork.Rtc.Sdk.Provider.Janus", "providers/Sdkwork.Rtc.Sdk.Provider.Janus/Sdkwork.Rtc.Sdk.Provider.Janus.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Janus/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Janus/src/RtcProviderJanusPackageContract.cs", "RtcProviderJanusPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
        new("mediasoup", "rtc-mediasoup", "sdkwork-rtc-driver-mediasoup", "Sdkwork.Rtc.Sdk.Provider.Mediasoup", "providers/Sdkwork.Rtc.Sdk.Provider.Mediasoup/Sdkwork.Rtc.Sdk.Provider.Mediasoup.csproj", "providers/Sdkwork.Rtc.Sdk.Provider.Mediasoup/README.md", "providers/Sdkwork.Rtc.Sdk.Provider.Mediasoup/src/RtcProviderMediasoupPackageContract.cs", "RtcProviderMediasoupPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
    ];

public static RtcProviderPackageCatalogEntry? GetRtcProviderPackageByProviderKey(string providerKey) =>
        Entries.FirstOrDefault(entry => entry.providerKey == providerKey);

    public static RtcProviderPackageCatalogEntry? GetRtcProviderPackageByPackageIdentity(string packageIdentity) =>
        Entries.FirstOrDefault(entry => entry.packageIdentity == packageIdentity);

}

File: sdkwork-rtc-sdk-java/src/main/java/com/sdkwork/rtc/metadata/RtcProviderPackageCatalog.java
package com.sdkwork.rtc.metadata;

import java.util.List;
import java.util.Optional;

public final class RtcProviderPackageCatalog {

  public static final List<RtcProviderPackageCatalogEntry> ENTRIES = List.of(
      new RtcProviderPackageCatalogEntry("volcengine", "rtc-volcengine", "sdkwork-rtc-driver-volcengine", "com.sdkwork:rtc-sdk-provider-volcengine", "providers/rtc-sdk-provider-volcengine/pom.xml", "providers/rtc-sdk-provider-volcengine/README.md", "providers/rtc-sdk-provider-volcengine/src/main/java/com/sdkwork/rtc/provider/volcengine/RtcProviderVolcenginePackageContract.java", "RtcProviderVolcenginePackageContract", true, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("aliyun", "rtc-aliyun", "sdkwork-rtc-driver-aliyun", "com.sdkwork:rtc-sdk-provider-aliyun", "providers/rtc-sdk-provider-aliyun/pom.xml", "providers/rtc-sdk-provider-aliyun/README.md", "providers/rtc-sdk-provider-aliyun/src/main/java/com/sdkwork/rtc/provider/aliyun/RtcProviderAliyunPackageContract.java", "RtcProviderAliyunPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("tencent", "rtc-tencent", "sdkwork-rtc-driver-tencent", "com.sdkwork:rtc-sdk-provider-tencent", "providers/rtc-sdk-provider-tencent/pom.xml", "providers/rtc-sdk-provider-tencent/README.md", "providers/rtc-sdk-provider-tencent/src/main/java/com/sdkwork/rtc/provider/tencent/RtcProviderTencentPackageContract.java", "RtcProviderTencentPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("agora", "rtc-agora", "sdkwork-rtc-driver-agora", "com.sdkwork:rtc-sdk-provider-agora", "providers/rtc-sdk-provider-agora/pom.xml", "providers/rtc-sdk-provider-agora/README.md", "providers/rtc-sdk-provider-agora/src/main/java/com/sdkwork/rtc/provider/agora/RtcProviderAgoraPackageContract.java", "RtcProviderAgoraPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("zego", "rtc-zego", "sdkwork-rtc-driver-zego", "com.sdkwork:rtc-sdk-provider-zego", "providers/rtc-sdk-provider-zego/pom.xml", "providers/rtc-sdk-provider-zego/README.md", "providers/rtc-sdk-provider-zego/src/main/java/com/sdkwork/rtc/provider/zego/RtcProviderZegoPackageContract.java", "RtcProviderZegoPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("livekit", "rtc-livekit", "sdkwork-rtc-driver-livekit", "com.sdkwork:rtc-sdk-provider-livekit", "providers/rtc-sdk-provider-livekit/pom.xml", "providers/rtc-sdk-provider-livekit/README.md", "providers/rtc-sdk-provider-livekit/src/main/java/com/sdkwork/rtc/provider/livekit/RtcProviderLivekitPackageContract.java", "RtcProviderLivekitPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("twilio", "rtc-twilio", "sdkwork-rtc-driver-twilio", "com.sdkwork:rtc-sdk-provider-twilio", "providers/rtc-sdk-provider-twilio/pom.xml", "providers/rtc-sdk-provider-twilio/README.md", "providers/rtc-sdk-provider-twilio/src/main/java/com/sdkwork/rtc/provider/twilio/RtcProviderTwilioPackageContract.java", "RtcProviderTwilioPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("jitsi", "rtc-jitsi", "sdkwork-rtc-driver-jitsi", "com.sdkwork:rtc-sdk-provider-jitsi", "providers/rtc-sdk-provider-jitsi/pom.xml", "providers/rtc-sdk-provider-jitsi/README.md", "providers/rtc-sdk-provider-jitsi/src/main/java/com/sdkwork/rtc/provider/jitsi/RtcProviderJitsiPackageContract.java", "RtcProviderJitsiPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("janus", "rtc-janus", "sdkwork-rtc-driver-janus", "com.sdkwork:rtc-sdk-provider-janus", "providers/rtc-sdk-provider-janus/pom.xml", "providers/rtc-sdk-provider-janus/README.md", "providers/rtc-sdk-provider-janus/src/main/java/com/sdkwork/rtc/provider/janus/RtcProviderJanusPackageContract.java", "RtcProviderJanusPackageContract", false, false, "future-runtime-bridge-only", "reserved"),
      new RtcProviderPackageCatalogEntry("mediasoup", "rtc-mediasoup", "sdkwork-rtc-driver-mediasoup", "com.sdkwork:rtc-sdk-provider-mediasoup", "providers/rtc-sdk-provider-mediasoup/pom.xml", "providers/rtc-sdk-provider-mediasoup/README.md", "providers/rtc-sdk-provider-mediasoup/src/main/java/com/sdkwork/rtc/provider/mediasoup/RtcProviderMediasoupPackageContract.java", "RtcProviderMediasoupPackageContract", false, false, "future-runtime-bridge-only", "reserved")
  );

public static Optional<RtcProviderPackageCatalogEntry> getRtcProviderPackageByProviderKey(String providerKey) {
    for (var entry : ENTRIES) {
      if (entry.providerKey().equals(providerKey)) {
        return Optional.of(entry);
      }
    }

    return Optional.empty();
  }

  public static Optional<RtcProviderPackageCatalogEntry> getRtcProviderPackageByPackageIdentity(String packageIdentity) {
    for (var entry : ENTRIES) {
      if (entry.packageIdentity().equals(packageIdentity)) {
        return Optional.of(entry);
      }
    }

    return Optional.empty();
  }


  private RtcProviderPackageCatalog() {
  }

  public record RtcProviderPackageCatalogEntry(
      String providerKey,
      String pluginId,
      String driverId,
      String packageIdentity,
      String manifestPath,
      String readmePath,
      String sourcePath,
      String sourceSymbol,
      boolean builtin,
      boolean rootPublic,
      String status,
      String runtimeBridgeStatus
  ) {
  }
}
